go-controller/pkg/config/config.go

```diff
@@ -757,54 +774,60 @@ func setOVSExternalID(exec kexec.Interface, key, value string) error {
 
 func buildKubernetesConfig(exec kexec.Interface, cli, file *config, saPath string, defaults *Defaults) error {
 	// token adn ca.crt may be from files mounted in container.
-	var saConfig KubernetesConfig
+	saConfig := savedKubernetes
 	if data, err := ioutil.ReadFile(filepath.Join(saPath, kubeServiceAccountFileToken)); err == nil {
 		saConfig.Token = string(data)
 	}
 	if _, err2 := os.Stat(filepath.Join(saPath, kubeServiceAccountFileCACert)); err2 == nil {
 		saConfig.CACert = filepath.Join(saPath, kubeServiceAccountFileCACert)
 	}
 
-	if err := overrideFields(&Kubernetes, &saConfig); err != nil {
+	if err := overrideFields(&Kubernetes, &saConfig, &savedKubernetes); err != nil {
 		return err
 	}
 
-	// Grab default values from OVS external IDs
-	if defaults.K8sAPIServer {
-		Kubernetes.APIServer = getOVSExternalID(exec, "k8s-api-server")
-	}
-	if defaults.K8sToken {
-		Kubernetes.Token = getOVSExternalID(exec, "k8s-api-token")
-	}
-	if defaults.K8sCert {
-		Kubernetes.CACert = getOVSExternalID(exec, "k8s-ca-certificate")
-	}
-
 	// values for token, cacert, kubeconfig, api-server may be found in several places.
-	// Take the first found when looking in this order: command line options, config file,
+	// Priority order (highest first): OVS config, command line options, config file,
 	// environment variables, service account files
 
-	envConfig := KubernetesConfig{
-		Kubeconfig: os.Getenv("KUBECONFIG"),
-		CACert:     os.Getenv("K8S_CACERT"),
-		APIServer:  os.Getenv("K8S_APISERVER"),
-		Token:      os.Getenv("K8S_TOKEN"),
+	envConfig := savedKubernetes
+	envVarsMap := map[string]string{
+		"Kubeconfig": "KUBECONFIG",
+		"CACert":     "K8S_CACERT",
+		"APIServer":  "K8S_APISERVER",
+		"Token":      "K8S_TOKEN",
+	}
+	for k, v := range envVarsMap {
+		if x, exists := os.LookupEnv(v); exists && len(x) > 0 {
+			reflect.ValueOf(&envConfig).Elem().FieldByName(k).SetString(x)
+		}
 	}
 
-	if err := overrideFields(&Kubernetes, &envConfig); err != nil {
+	if err := overrideFields(&Kubernetes, &envConfig, &savedKubernetes); err != nil {
 		return err
 	}
 
 	// Copy config file values over default values
-	if err := overrideFields(&Kubernetes, &file.Kubernetes); err != nil {
+	if err := overrideFields(&Kubernetes, &file.Kubernetes, &savedKubernetes); err != nil {
 		return err
 	}
 
 	// And CLI overrides over config file and default values
-	if err := overrideFields(&Kubernetes, &cli.Kubernetes); err != nil {
+	if err := overrideFields(&Kubernetes, &cli.Kubernetes, &savedKubernetes); err != nil {
 		return err
 	}
 
+	// Grab default values from OVS external IDs
```

Comment:
Note I moved this down here, so that if the behavior is if the booleans are set to detect from OVS, they will override all other settings. Not sure if that is the desired behavior or not.

Reply:
@girishmg will this be OK with you? We don't use the OVS external IDs, so I think this would only affect your setups.